modules/keyvault/secret-autorotation/main.bicep

```
@@ -0,0 +1,348 @@
+}]
+
+@description('Create access policy of the keyvault secrets for the function app')
+module addAccessPolicy './modules/access-policy.bicep' = [for item in items(reduce(secrets, {}, (cur, next) => union(cur, { '${next.keyvaultName}': (contains(next, 'keyvaultRg') ? next.keyvaultRg : resourceGroup().name) }))): {
+  name: 'ap-${uniqueString(item.key)}-${suffix}'
+  scope: resourceGroup(item.value)
+  params: {
+    keyvaultName: item.key
+    principalId: functionAppIdentityType == 'SystemAssigned' ? functions.identity.principalId : uai.properties.principalId
+  }
+  dependsOn: [ deployFunctionZip ]
+}]
+
+@batchSize(1)
+@description('Create Keyvault event subscription and let the function app consume it.')
+module eventSubscription './modules/event-subscription.bicep' = [for item in secrets: if (systemTopicName == '') {
+  name: 'kv-event-sub-${uniqueString('${item.keyvaultName}${item.secretName}')}-${suffix}'
+  scope: resourceGroup(contains(item, 'keyvaultRg') ? item.keyvaultRg : resourceGroup().name)
+  params: {
+    functionAppName: functionAppName
+    functionAppRg: resourceGroup().name
+    functionName: rotationFunctionNames[item.type]
+    keyvaultName: item.keyvaultName
+    secretName: item.secretName
+  }
+  dependsOn: [
+    grantAccess
+    addAccessPolicy
+  ]
+}]
+
+@batchSize(1)
+@description('Create event subscription to the system topic and let the function app consume it.')
+module topicSubscription './modules/topic-subscription.bicep' = [for item in secrets: if (systemTopicName != '') {
+  name: 'topic-sub-${uniqueString('${item.keyvaultName}${item.secretName}')}-${suffix}'
+  scope: resourceGroup(systemTopicRg)
+  params: {
+    functionAppName: functionAppName
+    functionAppRg: resourceGroup().name
+    functionName: rotationFunctionNames[item.type]
+    keyvaultName: item.keyvaultName
+    secretName: item.secretName
+    systemTopicName: systemTopicName
+  }
+  dependsOn: [ deployFunctionZip ]
+}]
+
+@batchSize(1)
+@description('Create keyvault secret if needed. Primary key of the target resource will be used as the initial secret value.')
+module vaultSecrets './modules/create-secret.bicep' = [for item in secrets: if (contains(item, 'isCreate') && item.isCreate) {
+  name: 'kv-secret-${uniqueString('${item.keyvaultName}${item.secretName}')}-${suffix}'
+  scope: resourceGroup(contains(item, 'keyvaultRg') ? item.keyvaultRg : resourceGroup().name)
+  params: {
+    type: item.type
+    resourceName: item.resourceName
+    resourceRg: contains(item, 'resourceRg') ? item.resourceRg : resourceGroup().name
+    keyvaultName: item.keyvaultName
+    secretName: item.secretName
+    validityDays: contains(item, 'validityDays') ? item.validityDays : 90
+    expirationDate: contains(item, 'expirationDate') ? item.expirationDate : dateTimeAdd(now, 'P${contains(item, 'validityDays') ? item.validityDays : 90}D')
+  }
+}]
+
+@description('ID of the function app created.')
+output siteId string = functions.id
```

```suggestion
output id string = functions.id
```

I like to name the output id of primary resource in my template simply 'id' to make it like basic azure resources.
Also consider outputting the ids of the other resourced created. Even if in an object like 

```bicep
nestedResource object string = {
  appInsightsId: appInsights.id
  ...
}

```
I also like to output the names to make it easy to create pointers using `existing`.